src/MSBuild/XMake.cs

```diff
@@ -1240,6 +1248,35 @@ private static BuildResult ExecuteBuild(BuildManager buildManager, BuildRequestD
             return s_activeBuild.Execute();
         }
 
+        private static BuildResult ExecuteGraphBuild(
+            string projectFile,
+            string[] targets,
+            BuildManager buildManager,
+            ProjectCollection projectCollection,
+            Dictionary<string, string> globalProperties,
+            string toolsVersion)
+        {
+            lock (s_buildLock)
+            {
+                var projectGraph = new ProjectGraph(projectFile, projectCollection, globalProperties, toolsVersion);
+
+                // TODO: Do a full graph traversal
+                var projectInstance = projectGraph.EntryProjectNode.Project.CreateProjectInstance();
+                var request = new BuildRequestData(projectInstance, targets);
+
+                s_activeBuild = buildManager.PendBuildRequest(request);
```

Comment:
This is extremely rudimentary for now. I think using the build manager as a black box is an OK initial approach, but I think a better eventual solution is to just spam the build manager with requests for each project providing the correct "blocker" data to describe the dependency graph correctly. The build manager already has a JIT dependency graph effectively for who is blocked on who, but unfortunately as it stands, it's "which node is blocked on which request", while we need more of "which request is blocked on which request". Anyway, something to explore more when we get there.

Reply:
I agree this is fine for now. I'm not sure what the right thing will be in the end.

Reply:
Options David and I discussed:
- for quick prototyping, use the BuildManager as a black box. The graph builder instantiates a build manager, walks the projects bottom up and keeps pending builds on the BuildManager.
- scheduler based
  - the Scheduler already constructs a JIT dependency graph. If we can adapt the static graph to Scheduler graph, it might just work as expected. I like this option because it reuses the Scheduler mostly as is.
  - the Scheduler implements IScheduler. We can add another implementation. This is nice if we think we can do better than the Scheduler. It's also nice that we'd hopefully make something more readable / understandable. Downside is that we'll need understand more details about current scheduling.

Reply:
I'll open a work item to investigate proper scheduling.